decision_service/dsclientlib/ds_driver.cc

```diff
@@ -0,0 +1,131 @@
+#include "ds_driver.h"
+#include "ds_event.h"
+#include "logger/ds_logger.h"
+
+#include <boost/uuid/uuid_generators.hpp> // generators
+#include <boost/uuid/uuid_io.hpp>         // streaming operators etc.
+#include "error_callback_fn.h"
+
+
+//this macro assumes that success_code equals 0
+#define TRY_OR_RETURN(x) do { \
+  int retval = (x); \
+  if (retval != 0) { \
+    return retval; \
+  } \
+} while (0)
+
+namespace decision_service {
+
+	int check_null_or_empty(const char * arg1, const char * arg2, api_status* status);
+
+	class driver_impl {
+	public:
+    	using error_fn = void(*)(const api_status&,void* user_context);
+
+	private:
+	  	utility::config_collection _configuration;
+    	error_callback_fn _error_cb;
+    	logger _logger;
+
+	public:
+		int ranking_request(const char * uuid, const char * context, ranking_response & response, api_status * status)
+		{
+	    	//clear previous errors if any
+      		if (status != nullptr) status->clear();
+
+			//check arguments
+			TRY_OR_RETURN(check_null_or_empty(uuid, context, status));
+
+			/* GET ACTIONS PROBABILITIES FROM VW */
+
+			/**/ // TODO: replace with call to parse example and predict
+			/**/ // TODO: once that is complete
+				/**/ std::vector<std::pair<int, float>> action_proba;
+			/**/ std::string model_id = "model_id";
+			/**/ action_proba.push_back(std::pair<int, float>(2, 0.4f));
+			/**/ action_proba.push_back(std::pair<int, float>(1, 0.3f));
+			/**/ action_proba.push_back(std::pair<int, float>(4, 0.2f));
+			/**/ action_proba.push_back(std::pair<int, float>(3, 0.1f));
+
+			//send the ranking event to the backend
+			ranking_event evt(uuid, context, action_proba, model_id);
+			TRY_OR_RETURN(_logger.append_ranking(evt.serialize(), status));
+
+			response.set_uuid(uuid);
+			response.set_ranking(action_proba);
+
+			return error_code::success;
+		}
+
+		//here the uuid is auto-generated
+		int ranking_request(const char * context, ranking_response & response, api_status * status)
```

Comment:
>const char * context, ranking_response & response, api_status * status [](start = 22, length = 70)

can we have consistent code style const char * vs const char* (space between *)

Reply:
done

---
In reply to: [186735496](https://github.com/JohnLangford/vowpal_wabbit/pull/1479#discussion_r186735496) [](ancestors = 186735496)